src/Compilers/Test/Core/Assert/AssertEx.cs

```diff
@@ -594,37 +616,117 @@ public static void AssertResultsEqual(string result1, string result2)
 
         public static void AssertContainsToleratingWhitespaceDifferences(string expectedSubString, string actualString)
         {
-            expectedSubString = NormalizeWhitespace(expectedSubString);
-            actualString = NormalizeWhitespace(actualString);
-            Assert.Contains(expectedSubString, actualString, StringComparison.Ordinal);
+            AssertContainsToleratingWhitespaceDifferences(expectedSubString.AsSpan(), actualString.AsSpan());
+        }
+
+        public static void AssertContainsToleratingWhitespaceDifferences(ReadOnlySpan<char> expectedSubString, ReadOnlySpan<char> actualString)
+        {
+            expectedSubString = NormalizeWhitespace(expectedSubString, out var pooled1);
+            actualString = NormalizeWhitespace(actualString, out var pooled2);
+
+            if (!actualString.Contains(expectedSubString, StringComparison.Ordinal))
+            {
+                // Call through to actual API to get same error message experience
+                Assert.Contains(expectedSubString.ToString(), actualString.ToString(), StringComparison.Ordinal);
+            }
+
+            if (pooled1 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled1);
+            }
+            if (pooled2 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled2);
+            }
         }
 
         public static void AssertStartsWithToleratingWhitespaceDifferences(string expectedSubString, string actualString)
         {
-            expectedSubString = NormalizeWhitespace(expectedSubString);
-            actualString = NormalizeWhitespace(actualString);
-            Assert.StartsWith(expectedSubString, actualString, StringComparison.Ordinal);
+            AssertStartsWithToleratingWhitespaceDifferences(expectedSubString.AsSpan(), actualString.AsSpan());
+        }
+
+        public static void AssertStartsWithToleratingWhitespaceDifferences(ReadOnlySpan<char> expectedSubString, ReadOnlySpan<char> actualString)
+        {
+            expectedSubString = NormalizeWhitespace(expectedSubString, out var pooled1);
+            actualString = NormalizeWhitespace(actualString, out var pooled2);
+
+            if (!actualString.StartsWith(expectedSubString, StringComparison.Ordinal))
+            {
+                // Call through to actual API to get same error message experience
+                Assert.StartsWith(expectedSubString.ToString(), actualString.ToString(), StringComparison.Ordinal);
+            }
+
+            if (pooled1 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled1);
+            }
```

Comment:
I understand this pattern but it's also feels a bit much to me. Too much manual work. I'm wondering if we should instead approach it with something like the following: 

```csharp
struct ArrayPoolValue<T> : IDisposable
{
  private T[] _array;
  public T[] T Array
  {
    get => _array;
    set
    {
       if (_array is not null)
        ArrayPool<T>.Shared.Return(_array);
       _array = value;
    }
  }
  public void Dispose()
  {
    if (_array is not null)
    {
        ArrayPool<T>.Shared.Return(_array);
        _array = null;
    }
  }
}
```

Then we could change the signatures  to take `ref ArrayPoolValue<T> pool`. Consumption becomes simply: 

```csharp
using var pooled1 = new ArrayPool<char>;
```

@333fred for thoughts

Reply:
`MemoryPool<T>` and `IMemoryOwner<T>` does exactly that, but with overhead of extra abstraction through `Memory<T>`.

Reply:
I don't think I care that much. These helpers are presumably not touched much, and I don't think we care about ensuring it's always returned?

Reply:
> and I don't think we care about ensuring it's always returned

No, it's not a correctness issue to not return it. Unless assertion failure is expected (e.g., if you have a bunch of tests like `bool didError = false; try { ... calls Assert(); } catch { didError = true; } Assert.True(didError);`), there's not really any reason to worry about ensuring it imo.

> Then we could change the signatures to take `ref ArrayPoolValue<T>` pool. Consumption becomes simply:
> `using var pooled1 = new ArrayPool<char>;`

You can't pass `using` variables by reference, we will very quickly run into this issue I think.